core/src/subgraph/loader.rs

```diff
@@ -0,0 +1,247 @@
+        // Turn the entities into unresolved data sources
+        future::ok(
+            entities
+                .into_iter()
+                .map(Into::into)
+                .collect::<Vec<UnresolvedDataSource>>(),
+        )
+    }
+
+    fn resolve_data_sources(
+        self: Arc<Self>,
+        unresolved_data_sources: Vec<UnresolvedDataSource>,
+    ) -> impl Future<Item = Vec<DataSource>, Error = Error> + Send {
+        // Resolve the data sources and return them
+        stream::iter_ok(unresolved_data_sources).fold(vec![], move |mut resolved, data_source| {
+            data_source
+                .resolve(&*self.link_resolver)
+                .and_then(|data_source| {
+                    resolved.push(data_source);
+                    future::ok(resolved)
+                })
+        })
+    }
+}
+
+impl<L, Q, S> DataSourceLoaderTrait for DataSourceLoader<L, Q, S>
+where
+    L: LinkResolver,
+    Q: GraphQlRunner,
+    S: Store + SubgraphDeploymentStore,
+{
+    fn load_dynamic_data_sources(
+        self: Arc<Self>,
+        deployment_id: &SubgraphDeploymentId,
+    ) -> Box<Future<Item = Vec<DataSource>, Error = Error> + Send> {
+        struct LoopState {
+            data_sources: Vec<DataSource>,
+            skip: i32,
+        }
+
+        let initial_state = LoopState {
+            data_sources: vec![],
+            skip: 0,
+        };
+
+        // Clones for async looping
+        let self1 = self.clone();
+        let deployment_id = deployment_id.clone();
+
+        Box::new(
+            future::loop_fn(initial_state, move |mut state| {
+                let deployment_id1 = deployment_id.clone();
+                let deployment_id2 = deployment_id.clone();
+                let deployment_id3 = deployment_id.clone();
+
+                let self2 = self1.clone();
+                let self3 = self1.clone();
+                let self4 = self1.clone();
+                let self5 = self1.clone();
+                let self6 = self1.clone();
+
+                future::result(self2.dynamic_data_sources_query(&deployment_id1, state.skip))
+                    .and_then(move |query| self3.query_dynamic_data_sources(deployment_id2, query))
+                    .and_then(move |query_result| {
+                        self4.parse_data_sources(deployment_id3, query_result)
+                    })
+                    .and_then(move |typed_entities| {
+                        self5.convert_to_unresolved_data_sources(typed_entities)
+                    })
+                    .and_then(move |unresolved_data_sources| {
+                        self6.resolve_data_sources(unresolved_data_sources)
+                    })
+                    .map(move |data_sources| {
+                        if data_sources.is_empty() {
+                            future::Loop::Break(state)
+                        } else {
+                            state.skip += dbg!(data_sources.len() as i32);
```

Comment:
That's why it _is_ awesome at all, because you can plug it in anywhere and the value is both printed and passed through.

Reply:
Sure it's awesome for debugging, but do we want to print anything without going through a logger? Afaik this can potentially get garbled with log messages.

Reply:
Oh, yeah, I wasn't arguing it should remain there. 😆 In fact, I thought I had removed it already. Done now.